### ISSUE TITLE
cmd/vet: check for improper use of t.Run and t.Parallel

### ISSUE THREAD

**Issue by dsnet** [golang/go#18106]
Consider the following test:
```go
func Test(t *testing.T) {
	for _, name := range []string{"foo", "bar", "baz"} {
		data := name // This line is necessary
		t.Run(name, func(t *testing.T) {
			t.Parallel()
			t.Log(data)
		})
	}
}
```

Without the `name := name` line to capture the name variable, the subtest will execute with the wrong values as the testdata. Worst yet, the tests will run with the correct names, but the wrong data, completely misleading the user that the test is executing as expected.

The [blog on subtests](https://blog.golang.org/subtests) explicitly warns that this must be done:
> Note that we need to capture the range variable to ensure that tc gets bound to the correct instance.

I propose that vet check for this mis-use. One possible algorithm:
* If `t.Parallel` is called within a `t.Run`
* and the body of `t.Run` following the `t.Parallel` references a variable scoped to the `range` of some outer loop (i.e., not captured)
* then flag as potential issue.

\cc @dominikh

**Comment by quentinmit** [comment]
What is the frequency with which this occurs?

**Comment by minux** [comment]
this is just one possible manifest of
https://golang.org/doc/faq#closures_and_goroutines. If we can do a generic
vet check for all occurrences, then it's fine. But I don't think checking
only this single case is appropriate.

**Comment by gonzaloserrano** [comment]
For the record it just happened to us at Social Point. Maybe a rule in https://github.com/dominikh/go-staticcheck would make it too if in vet is not possible. cc/ @dominikh

**Comment by rsc** [comment]
#16520 is the generalization of this. This is marked as NeedsDecision but really it needs a concrete proposal of what gets checked and how, and an argument for why that will have approximately no false positives.

**Comment by findleyr** [comment]
Just found this via search. For the record, this landed as #55972.
